h2o-automl/src/main/java/ai/h2o/automl/TrainingStep.java

```diff
@@ -0,0 +1,339 @@
+package ai.h2o.automl;
+
+import ai.h2o.automl.WorkAllocations.JobType;
+import ai.h2o.automl.WorkAllocations.Work;
+import hex.Model;
+import hex.ModelBuilder;
+import hex.ScoreKeeper;
+import hex.ensemble.StackedEnsembleModel;
+import hex.grid.Grid;
+import hex.grid.GridSearch;
+import hex.grid.HyperSpaceSearchCriteria;
+import water.Iced;
+import water.Job;
+import water.Key;
+import water.exceptions.H2OIllegalArgumentException;
+import water.util.Log;
+
+import java.util.Map;
+
+/**
+ * Parent class defining common properties and common logic for actual {@link AutoML} training steps.
+ */
+public abstract class TrainingStep<M extends Model> extends Iced<TrainingStep> {
+
+    private transient AutoML _aml;
+
+    protected Algo _algo;
+    protected String _id;
+    protected int _weight;
+    protected boolean _ignoreConstraints;  // whether or not to ignore the max_models/max_runtime constraints
+    protected String _description;
+
+    StepDefinition _fromDef;
+
+    protected TrainingStep(Algo algo, String id, int weight, AutoML autoML) {
+        _algo = algo;
+        _id = id;
+        _weight = weight;
+        _aml = autoML;
+        _description = algo.name()+" "+id;
+    }
+
+    protected abstract Work getWork();
```

Comment:
I should rename this to `getAllocatedWork` to avoid confusion with `makeWork` as they're not used during the same phase.